### PR TITLE
[LLVM] Add `__builtin_readsteadycounter` intrinsic and builtin for realtime clocks

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
This patch adds a new intrinsic and builtin function mirroring the
existing `__builtin_readcyclecounter`. The difference is that this
implementation targets a separate counter that some targets have which
returns a fixed frequency clock that can be used to determine elapsed
time, this is different compared to the cycle counter which often has
variable frequency.

This patch only adds support for the NVPTX and AMDGPU targets.

This is done as a new and separate builtin rather than an argument to
`readcyclecounter` to avoid needing to change existing code and to make
the separation more explicit.

### CODE REVIEW

**Comment by jhuber6** [comment]
This is a draft, I'm trying to get this to map to either `s_memtime` or `s_sendmsg_rtn(0x83)` depending on the target for AMDGPU. Currently, it will recognize the new intrinsic but just lower it to `i64 0`. I'm not overly familiar with the backend, so any suggestions would be greatly appreciated.

**Comment by jhuber6** [comment]
Formatting is expected to fail to preserve local style.

File: llvm/lib/CodeGen/IntrinsicLowering.cpp [hunk 1]
```diff
@@ -312,6 +312,12 @@ void IntrinsicLowering::LowerIntrinsicCall(CallInst *CI) {
     CI->replaceAllUsesWith(ConstantInt::get(Type::getInt64Ty(Context), 0));
     break;
   }
+  case Intrinsic::readfixedtimer: {
+    errs() << "WARNING: this target does not support the llvm.readfixedtimer"
+           << " intrinsic.  It is being lowered to a constant 0\n";
+    CI->replaceAllUsesWith(ConstantInt::get(Type::getInt64Ty(Context), 0));
```
Comment (shiltian) [inline llvm/lib/CodeGen/IntrinsicLowering.cpp:318]: `Constant::getNullValue(Type::getInt64Ty(Context))`?

**Review by shiltian — COMMENTED** [review]
Generally looks good to me. Just not sure about the name. "fixed timer" sounds pretty confusing to me. probably `readfixedfreqtimer`?

File: llvm/lib/CodeGen/IntrinsicLowering.cpp (hunk 1 above)
Comment (jhuber6) [reply llvm/lib/CodeGen/IntrinsicLowering.cpp:318]: I copied it from the other one, but either way works. I think setting it to zero is a little more explicit.

**Comment by jhuber6** [comment]
> Generally looks good to me. Just not sure about the name. "fixed timer" sounds pretty confusing to me. probably `readfixedfreqtimer`?

Naming is the hard part. I was also thinking about `readrealtimecounter` or something. Maybe `readsteadycounter`?

File: llvm/lib/CodeGen/IntrinsicLowering.cpp [hunk 2]
```diff
@@ -312,6 +312,12 @@ void IntrinsicLowering::LowerIntrinsicCall(CallInst *CI) {
     CI->replaceAllUsesWith(ConstantInt::get(Type::getInt64Ty(Context), 0));
     break;
   }
+  case Intrinsic::readfixedtimer: {
+    errs() << "WARNING: this target does not support the llvm.readfixedtimer"
```
Comment (arsenm) [inline llvm/lib/CodeGen/IntrinsicLowering.cpp:316]: I'm surprised we have this random printing here. Is IntrinicLowering even used?

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td [hunk 3]
```diff
@@ -3759,7 +3759,6 @@ def CALL_PROTOTYPE :
 
 include "NVPTXIntrinsics.td"
 
-
```
Comment (arsenm) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3762]: Random whitespace change

**Review by arsenm — COMMENTED** [review]
Missing the clang builtin test

**Comment by MaskRay** [comment]
Looks good.

> LLVM] Add

Missing `[`?

**Comment by jhuber6** [comment]
Added clang test and renamed to `readsteadycounter` as I think it's more descriptive and matches the existing `readcyclecounter` better.

**Comment by RKSimon** [comment]
Are we assuming any particular relationship to __builtin_readcyclecounter in terms of scales etc? 

__builtin_readsteadycounter could be used to access x86 MPERF clock counters, but to access the corresponding APERF clock we'd then need a __builtin_readvariablecounter equivalent (__builtin_readcyclecounter gives the separate RDTSC clock value)

**Comment by jhuber6** [comment]
> Are we assuming any particular relationship to __builtin_readcyclecounter in terms of scales etc?
> 
> __builtin_readsteadycounter could be used to access x86 MPERF clock counters, but to access the corresponding APERF clock we'd then need a __builtin_readvariablecounter equivalent (__builtin_readcyclecounter gives the separate RDTSC clock value)

Not currently at least. The idea is just to expose the counter that has a fixed tick rate, compared to the cycle counter which is variable.

**Comment by JonChesterfield** [comment]
New intrinsic sounds right - a constant frequency counter is a different thing to a variable frequency counter.

"Steady" implies unchanging, so I'd agree with `readfixedfreqtimer` or similar.

We can't have a ratio between the two counters since one changes frequency and one doesn't.

Does x64 have something that maps usefully onto a fixed frequency counter intrinsic?

File: clang/docs/LanguageExtensions.rst [hunk 4]
```diff
@@ -2764,6 +2764,37 @@ Query for this feature with ``__has_builtin(__builtin_readcyclecounter)``. Note
 that even if present, its use may depend on run-time privilege or other OS
 controlled state.
 
+``__builtin_readsteadycounter``
+------------------------------
+
+``__builtin_readsteadycounter`` is used to access the fixed frequency counter
+register (or a similar steady-rate clock) on those targets that support it.
+The function is similar to ``__builtin_readcyclecounter`` above except that the
+frequency is fixed, making it suitable for measuring elapsed time.
```
Comment (Artem-B) [inline clang/docs/LanguageExtensions.rst:2773]: Should we mention that we do not guarantee any particular frequency, just that it's stable and it's up to the user to figure out the actual frequency, if they need to.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGDumper.cpp [hunk 5]
```diff
@@ -104,6 +104,7 @@ std::string SDNode::getOperationName(const SelectionDAG *G) const {
   case ISD::ATOMIC_STORE:               return "AtomicStore";
   case ISD::PCMARKER:                   return "PCMarker";
   case ISD::READCYCLECOUNTER:           return "ReadCycleCounter";
+  case ISD::READSTEADYCOUNTER:             return "ReadFixedTimer";
```
Comment (Artem-B) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGDumper.cpp:107]: Should it be "ReadSteadyCounter" ? 

Also, whitespace/alignment looks off.

**Review by Artem-B — COMMENTED** [review]
LGTM with few nits for general and NVPTX parts.

File: clang/docs/LanguageExtensions.rst (hunk 4 above)
Comment (jhuber6) [reply clang/docs/LanguageExtensions.rst:2773]: Good point, done.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGDumper.cpp (hunk 5 above)
Comment (jhuber6) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGDumper.cpp]: Forgot to fix that when I renamed it, thanks for spotting it.

**Review by arsenm — COMMENTED** [review]
Add to release notes?

**Comment by jhuber6** [comment]
> Add to release notes?

Done

**Review by shiltian — APPROVED** [review]
LG

**Pull request merged**